pkg/chaosdaemon/util.go

```diff
@@ -112,6 +115,10 @@ func GetChildProcesses(ppid uint32) ([]uint32, error) {
 	}
 }
 
+func (s *DaemonServer) getLoggerFromContext(ctx context.Context) logr.Logger {
```

Comment:
Is there a direct way to let developers know what is inside the ctx?
I have tried to find whether a ctx from time_server.SetTimeOffset have the namespase , but found a lot of interface with Client.Get(ctx,...), I think it is too hard to know it.

Reply:
The developer has to know where the context comes from. If the context is created by grpc-server (for every request), it's equipped with the `namespacedName` (automatically, with the help of grpc server-side interceptor).

Reply:
I suggest to add a comment about where it comes from add what will happen.

Reply:
> Is there a direct way to let developers know what is inside the ctx?

We(the provider of module/lbrary `DaemonServer`) could not assume that more things are carried with `context.Context`, because it's out of the scope of `context.Context`. 

The current `log.EnrichLoggerWithContext` works in a "best-effort pattern", it would append more information when context brings more information and would not be panic if it doesn't, which looks great.

Reply:
> The current log.EnrichLoggerWithContext works in a "best-effort pattern", it would append more information when context brings more information and would not be panic if it doesn't, which looks great.

But in most of package , I wish they log with chaos name forever. Like in http-chaos I want to ensure that logger wiil log with "HTTP-Chaos"

Reply:
@Andrewmatilde There is no way, at least in golang, to ensure the content of context in the compile-time. You can never avoid a bad guy to pass `context.TODO()` or `context.Background()` to these functions. What we could do is to find the problem as early as possible (during the review, or when we find the log is incorrect) and fix them.

Reply:
Why not a concrete context type? For example, we defined a concrete context `Ctx`, which implements interface 
`context.Context`. For the functions that need special content, we can receive `*Ctx` in parameter; and for the functions that only need common context, we just pass `*Ctx` as `context.Context`

Reply:
@Hexilee Using a concrete context `Ctx` means we couldn't use `context.With*` group of functions, which will return a type different from the beginning `Ctx` (see [context.WithValue](https://cs.opensource.google/go/go/+/refs/tags/go1.17.7:src/context/context.go;l=523)), and I don't want to put this limitation on the Chaos Mesh codes.

It has nearly the same problem with `error`. We need a `context.As(ctx, &Ctx)` to extract a `Ctx` from a context with an arbitrary type (but it doesn't solve any problem, it could also fail if it doesn't have a `Ctx` inside).

Reply:
This suggestion is a little interesting to think further. If the golang can define such interface:

```go
type Context interface {
    func WithValue(self, key, val interface{}) Self;
}
```

Then it would be fine. But it cannot! (see the `Self` type, which is [not implemented](https://github.com/golang/go/issues/28254))

Reply:
I think following function also meets our requirements:

```go
func WithValue[C Context](ctx C, key, value interface{}) C {

}
```

Maybe we can consider it in go-1.18 or later versions.

Reply:
> I think following function also meets our requirements:
> 
> ```go
> func WithValue[C Context](ctx C, key, value interface{}) C {}
> ```
> 
> Maybe we can consider it in go-1.18 or later versions.

My fault, it cannot be implemented in the stdlib. How about the following code 😂.

```
type Context interface {
    func WithValue(ret *Context, key, val interface{});
}
```

Reply:
```go
type Context interface {
    WithValue(ret *Context, key, val interface{});
}
```

Then the beautiful pure function becomes an ugly dirty function that mutates the object itself 😿 .

Reply:
> > ```go
> > type Context interface {
> >    WithValue(ret *Context, key, val interface{});
> > }
> > ```
> 
> Then the beautiful pure function becomes an ugly dirty function that mutates the object itself 😿 .

There may be some misunderstanding. Here are my examples:

```go
func doSomething(ctx *Ctx) {
    var newCtx Ctx
    ctx.WithValue(&newCtx, "name", "foo")
    go someFunc(&newCtx)
}
```

I think it's an acceptable API in the C programming language 😂.

Reply:
> But in most of package , I wish they log with chaos name forever. Like in http-chaos I want to ensure that logger wiil log with "HTTP-Chaos"

IMO, if you want **always** log the "chaos name", it means the "chaos name" is the required parameter. Please change the function signature, and append the new parameter `chaosName string` explicity.

Reply:
> > But in most of package , I wish they log with chaos name forever. Like in http-chaos I want to ensure that logger wiil log with "HTTP-Chaos"
> 
> IMO, if you want to **always** log the "chaos name", it means the "chaos name" is the required parameter. Please change the function signature, and append the new parameter `chaosName string` explicity.

I agree with you.